index.js

```diff
@@ -0,0 +1,148 @@
+var Command = require('./command');
+var Logging = require('./logging');
+var Input = require('./input');
+var Util = require('./util');
+var config = require('./config.json');
+var devices = config.devices;
+
+// Regular expression for extracting adb property output
+var GETPROP_MATCHER = /^\[([\s\S]*?)]: \[([\s\S]*?)]\r?$/gm;
+
+/**
+ * API for interacting with devices and environments
+ * @param {String} [serial]
+ * @constructor
+ */
+var Device = function(serial) {
+  this.serial = serial;
+};
+
+/**
+ * Map of properties obtained from `getprop`
+ * @type {Object}
+ */
+Device.prototype.properties = null;
+
+/**
+ * Fetch and set the device serial, verifying that the device is connected
+ * @returns {Promise}
+ */
+Device.prototype.setSerialVerified = function() {
+  var device = this;
+  var serial = device.serial;
+
+  device.serial = null;
+
+  return new Command('adb devices')
+    .exec()
+    .then(function(output) {
+      var lines = output.split('\n').slice(1, -2);
+
+      if (!lines.length) {
+        throw new Error('No devices found');
+      }
+
+      lines.some(function(line) {
+        var currentSerial = line.split('\t')[0];
+
+        if (!currentSerial.length) {
+          throw new Error('Unable to determine serial of connected device');
+        }
+
+        if (!serial) {
+          device.serial = currentSerial;
+          return true;
+        }
+
+        if (currentSerial === serial) {
+          device.serial = serial;
+          return true;
+        }
+
+        return false;
+      });
```

Comment:
During the boot-up of a device, or if the device is attached but in a bad state, adb can actually return 'offline', e.g:

rwood@ubuntu:~/gaia$ adb devices
List of devices attached 
emulator-5554	offline

Should that be checked here, to ensure that it says "attached"? Otherwise there is a chance it may think the device is attached but it isn't.